chain/consensus/mir/validator.go

```diff
@@ -0,0 +1,244 @@
+// OpaqueNetAddr can contain GRPC or Libp2p addresses.
+//
+// Examples of validator strings:
+// 	- t1wpixt5mihkj75lfhrnaa6v56n27epvlgwparujy@/ip4/127.0.0.1/tcp/10000/p2p/12D3KooWJhKBXvytYgPCAaiRtiNLJNSFG5jreKDu2jiVpJetzvVJ
+// 	- t1wpixt5mihkj75lfhrnaa6v56n27epvlgwparujy@127.0.0.1:1000
+// FIXME: Consider using json serde for this to support multiple multiaddr for validators.
+func ValidatorFromString(input string) (Validator, error) {
+	parts := strings.Split(input, "@")
+	if len(parts) != 2 {
+		return Validator{}, fmt.Errorf("failed to parse validators string")
+	}
+	ID := parts[0]
+	opaqueNetAddr := parts[1]
+
+	a, err := addr.NewFromString(ID)
+	if err != nil {
+		return Validator{}, err
+	}
+	addr, err := multiaddr.NewMultiaddr(opaqueNetAddr)
+	if err != nil {
+		return Validator{}, err
+	}
+
+	return Validator{
+		Addr:    a,
+		NetAddr: addr.String(),
+	}, nil
+}
+
+func (v *Validator) String() string {
+	return fmt.Sprintf("%s@%s", v.Addr.String(), v.NetAddr)
+}
+
+// ValidatorsToString adds a validator subnet, address and network address into a string.
+func ValidatorsToString(validators []Validator) string {
+	var s string
+	for _, v := range validators {
+		s += fmt.Sprintf("%s,", v.String())
+	}
+	return strings.TrimSuffix(s, ",")
+}
+
+func SplitAndTrimEmpty(s, sep, cutset string) []string {
+	if s == "" {
+		return []string{}
+	}
+
+	spl := strings.Split(s, sep)
+	nonEmptyStrings := make([]string, 0, len(spl))
+
+	for i := 0; i < len(spl); i++ {
+		element := strings.Trim(spl[i], cutset)
+		if element != "" {
+			nonEmptyStrings = append(nonEmptyStrings, element)
+		}
+	}
+
+	return nonEmptyStrings
+}
+
+type ValidatorSet struct {
+	Validators []Validator
+}
+
+func NewValidatorSet(vals []Validator) *ValidatorSet {
+	return &ValidatorSet{Validators: vals}
+}
+
+func (set *ValidatorSet) Size() int {
+	return len(set.Validators)
+}
+
+func (set *ValidatorSet) Equal(o *ValidatorSet) bool {
+	if set == nil && o == nil {
+		return true
+	}
+	if set == nil || o == nil {
+		return true
+	}
+	if set.Size() != o.Size() {
+		return false
+	}
+	for i, v := range set.Validators {
+		if v != o.Validators[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func (set *ValidatorSet) Hash() ([]byte, error) {
+	var hs []byte
+	for _, v := range set.Validators {
+		b, err := v.Bytes()
+		if err != nil {
+			return nil, err
+		}
+		hs = append(hs, b...)
+	}
+	return cid.NewCidV0(u.Hash(hs)).Bytes(), nil
+}
+
+func (set *ValidatorSet) GetValidators() []Validator {
+	return set.Validators
+}
+
+func (set *ValidatorSet) HasValidatorWithID(id string) bool {
+	for _, v := range set.Validators {
+		if v.ID() == id {
+			return true
+		}
+	}
+	return false
+}
+
+// BlockMiner returns a miner assigned deterministically using round-robin for an epoch to assign a reward
+// according to the rules of original Filecoin consensus.
+func (set *ValidatorSet) BlockMiner(epoch abi.ChainEpoch) addr.Address {
```

Comment:
The epoch here is the Filecoin epoch, not the Mir epoch, right? Can we add a comment on that?

Reply:
Fixed